### PR TITLE
HPCC-15608: Method Level Authentication

### PR DESCRIPTION
**Pull request by gaojx**
Initial implementation: support through attribute in ESPMethod:
   ESPMethod [auth_feature("auth-tag")] myMethod;

Signed-off-by: Jiafu Gao jiafu.gao@lexisnexis.com

### CODE REVIEW

**Comment by gaojx** [comment]
@rpastrana please code review.

**Comment by richardkchapman** [comment]
Needs to target master, not 6.0.2

File: esp/logging/loggingmanager/loggingmanager.cpp [hunk 1]
```diff
@@ -116,6 +116,7 @@ bool CLoggingManager::updateLog(const char* option, IEspContext& espContext, IPr
         const char* userId = espContext.queryUserId();
         if (userId && *userId)
             espContextTree->addProp("UserName", userId);
+        espContextTree->addPropInt64("ResponseTime", (__int64)(msTick()-espContext.queryCreationTime())/1000);
```
Comment (rpastrana) [inline esp/logging/loggingmanager/loggingmanager.cpp]: This looks to be from an unrelated jira.

File: initfiles/componentfiles/configxml/mysqlserver.xsd [hunk 2]
```diff
@@ -116,12 +116,19 @@
                     </xs:appinfo>
                 </xs:annotation>
             </xs:attribute>
-            <xs:attribute name="computer" type="computerType" use="required">
+            <xs:attribute name="computer" type="computerType" use="optional">
```
Comment (rpastrana) [inline initfiles/componentfiles/configxml/mysqlserver.xsd]: This doesn't seem to be related to this issue.

File: esp/esdllib/esdl_transformer2.cpp [hunk 3]
```diff
@@ -1511,6 +1511,18 @@ int Esdl2Transformer::process(IEspContext &ctx, EsdlProcessMode mode, const char
     if (!mi)
         throw MakeStringException(-1, "ESDL - method '%s::%s'not found", service, method);
 
+    // check the auth_feature
```
Comment (rpastrana) [inline esp/esdllib/esdl_transformer2.cpp]: are the auth_features declared in the request structure, or the method structure?

File: esp/esdllib/esdl_transformer2.cpp [hunk 4]
```diff
@@ -1511,6 +1511,18 @@ int Esdl2Transformer::process(IEspContext &ctx, EsdlProcessMode mode, const char
...
+    if (mode==EsdlRequestMode)
+    {
+        const char* auth_feature = mi->queryMetaData("auth_feature");
```
Comment (rpastrana) [inline esp/esdllib/esdl_transformer2.cpp]: we're working to make an auth_feature required for all methods.

File: esp/esdllib/esdl_transformer2.cpp [hunk 5]
```diff
@@ -1511,6 +1511,18 @@ int Esdl2Transformer::process(IEspContext &ctx, EsdlProcessMode mode, const char
...
+        if (auth_feature)
+        {
+            SecAccessFlags access;
+            if (!ctx.authorizeFeature(auth_feature,access) || access!=SecAccess_Full)
```
Comment (rpastrana) [inline esp/esdllib/esdl_transformer2.cpp]: the access level should be configurable as well, here only full is being allowed.

**Comment by rpastrana** [comment]
@gaojx left a few comments. some unrelated changes need to go in their own jira/pullrequest.
Take a look at 
https://github.com/hpcc-systems/HPCC-Platform/compare/master...rpastrana:HPCC-15375-AutoESPFeatureAuth for naming convention used for this feature

**Comment by richardkchapman** [comment]
Needs to target master branch

**Comment by rpastrana** [comment]
@afishbeck please review-- this PR is closed, but Jiafu is going to re-issue.

**Comment by richardkchapman** [comment]
@gaojx Any progress on this?

**Comment by rpastrana** [comment]
@gaojx let's make sure this is in sync with this change: https://github.com/hpcc-systems/HPCC-Platform/pull/9063

File: esp/esdllib/esdl_transformer2.cpp [hunk 6]
```diff
@@ -1511,6 +1511,18 @@ int Esdl2Transformer::process(IEspContext &ctx, EsdlProcessMode mode, const char
     if (!mi)
         throw MakeStringException(-1, "ESDL - method '%s::%s'not found", service, method);
 
+    // check the auth_feature
+    if (mode==EsdlRequestMode)
+    {
+        const char* auth_feature = mi->queryMetaData("auth_feature");
+        if (auth_feature)
```
Comment (rpastrana) [inline esp/esdllib/esdl_transformer2.cpp]: if (auth_feature && *auth_feature)

File: esp/esdllib/esdl_transformer2.cpp (hunk 5 above)
Comment (rpastrana) [inline esp/esdllib/esdl_transformer2.cpp]: let's add syntax to allow developer to specify level something like auth_feature("featurename:level"), we could default to full.

File: esp/logging/loggingmanager/loggingmanager.cpp (hunk 1 above)
Comment (rpastrana) [inline esp/logging/loggingmanager/loggingmanager.cpp]: @gaojx is there a pull request for this change?

File: initfiles/componentfiles/configxml/mysqlserver.xsd (hunk 2 above)
Comment (rpastrana) [inline initfiles/componentfiles/configxml/mysqlserver.xsd]: @gaojx is this still needed?
